Starforge/Editor/Render/LevelRender.cs

```diff
@@ -0,0 +1,300 @@
+                Rooms.Remove(dr);
+                dr.Target.Dispose();
+            }
+            Editor.Camera.Update();
+        }
+
+        /// <summary>
+        /// Disposes of the LevelRender.
+        /// </summary>
+        public void Dispose() {
+            foreach (DrawableRoom room in Rooms) {
+                room.Target.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Gets a DrawableRoom based on a Room.
+        /// </summary>
+        /// <param name="room">The room to search for.</param>
+        /// <returns>The DrawableRoom of the given room.</returns>
+        public DrawableRoom GetRoom(Room room) {
+            return Rooms.Where(d => d.Room == room).FirstOrDefault();
+        }
+
+        /// <summary>
+        /// Renders the entire level.
+        /// </summary>
+        public void Render() {
+            // If AlwaysRerender is on, always rerender rooms.
+            if (TargetUsage == RenderTargetUsage.DiscardContents) {
+                foreach (DrawableRoom room in VisibleRooms) {
+                    RenderRoom(room);
+                }
+            }
+
+            foreach (DrawableRoom room in VisibleRooms) {
+                if (room.Dirty) RenderRoom(room, Menubar.RerenderFlags);
+            }
+
+            Engine.Instance.GraphicsDevice.SetRenderTarget(null);
+            Engine.Instance.GraphicsDevice.Clear(Settings.BackgroundColor);
+
+            Engine.Batch.Begin(
+                SpriteSortMode.Deferred,
+                BlendState.AlphaBlend,
+                SamplerState.PointClamp,
+                DepthStencilState.None,
+                null,
+                null,
+                Editor.Camera.Transform
+            );
+
+            // Render rooms
+            foreach (DrawableRoom room in VisibleRooms) {
+                if (!room.Target.IsDisposed) Engine.Batch.Draw(room.Target, new Vector2(room.Room.X, room.Room.Y), Color.White);
+            }
+
+            // Render selected room
+            if (SelectedRoom != null && !Overlay.IsDisposed) {
+                Engine.Batch.Draw(Overlay, new Vector2(SelectedRoom.Room.X, SelectedRoom.Room.Y), Color.White * 0.75f);
+            }
+
+            // Render fillers
+            foreach (Rectangle r in Level.Fillers) {
+                GFX.Draw.Rectangle(r, Color.DimGray);
+            }
+
+            Engine.Batch.End();
+        }
+
+        /// <summary>
+        /// Renders a room.
+        /// </summary>
+        /// <param name="room">The room to render.</param>
+        /// <param name="flags">The render flags.</param>
+        public void RenderRoom(DrawableRoom room, RenderFlags flags = RenderFlags.All) {
+            // Generate decals
+            room.BGDecals = CreateDecalTextureList(room.Room.BackgroundDecals);
+            room.FGDecals = CreateDecalTextureList(room.Room.ForegroundDecals);
+
+            // Generate tiles
+            if (room.TilesDirty) {
```

Comment:
I personally think it makes more sense to leave the Dirty check outside of the RenderRoom function, although there is room for debate on this. The function is used in 4 places:
- `LevelRender.Render` (x2)
- `MapEditor.LoadLevel` (prerenders level)
- `MapEditor.RerenderAll` (rerenders all rooms)

I think it makes more sense to leave the check outside of the `RenderRoom` function itself because then it makes it more clear that `LevelRender.Render` is only rerendering the rooms if they need to be, not calling the function regardless and then having it check there.